odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,546 @@
+
+    The template gradient ``grad(I)`` after deformation with displacement
+    field ``v``:
+
+        grad(I)(Id + v)
+
+    where ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement field:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain.
+
+    Parameters
+    ----------
+    grad_f: `ProductSpace` element
+        Gradient of the template, i.e. a vector field on the
+        image domain
+    displacement : `ProductSpace` element
+        The vector field (displacement field) used in the linearized
+        deformation.
+    Examples
+    --------
+    >>> import odl
+    >>> space = odl.uniform_discr(0, 1, 5)
+    >>> template = space.element([0, 0, 1, 0, 0])
+    >>> grad_op = odl.Gradient(template.space, method='forward')
+    >>> template_grad = grad_op(template)
+    >>> template_grad
+    ProductSpace(uniform_discr(0.0, 1.0, 5), 1).element([
+        [0.0, 5.0, -5.0, 0.0, 0.0]
+    ])
+    >>> disp_field_space = grad_op.range
+    >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+    >>> deform_grad(template_grad, displacement_field)
+    ProductSpace(uniform_discr(0.0, 1.0, 5), 1).element([
+        [0.0, 5.0, -5.0, -5.0, 0.0]
+    ])
+    """
+    return displacement.space.element(
+        [linear_deform(gf, displacement) for gf in grad_f])
+
+
+class LinDeformFixedTempl(Operator):
+
+    """Operator mapping a displacement field to corresponding deformed template.
+
+    The operator maps a displacement field to the corresponding
+    deformed fixed template ``I``:
+
+        v --> I(Id + v)
+
+    where ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement field:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain.
+    """
+
+    def __init__(self, template):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLp` element
+            Fixed template that is to be deformed by the displacement field.
+
+        Examples
+        --------
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5)
+        >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> displacement_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+        >>> op = LinDeformFixedTempl(template)
+        >>> op(displacement_field)
+        uniform_discr(0.0, 1.0, 5).element([0.0, 0.0, 1.0, 1.0, 0.0])
+        """
+        if not isinstance(template.space, DiscreteLp):
+            raise TypeError('`template.space` {!r} not an element of'
+                            '`DiscreteLp`'.format(template.space))
+
+        domain_space = ProductSpace(template.space, template.space.ndim)
+        range_space = template.space
+
+        Operator.__init__(self, domain_space, range_space, linear=False)
+
+        self.template = template
```

Comment:
Not sure we want to expose the template that prominently (especially with write access). I'd prefer `self._template` plus a property `template`.